File: src/main/java/io/vproxy/vfx/ui/table/VTableCellPane.java
package io.vproxy.vfx.ui.table;

import javafx.beans.value.ChangeListener;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.layout.Background;
import javafx.scene.layout.StackPane;

public class VTableCellPane<S> extends StackPane {
    private final VTableRow<S> row;
    final ChangeListener<? super Number> heightWatcher = (ob, old, now) -> {
        if (now == null) return;
        double dNow = now.doubleValue();
        var thisHeight = getHeight();
        if (thisHeight < dNow) {
            setPrefHeight(dNow);
        }
    };
    private Background backgroundSetByUser = null;

    public VTableCellPane(Node node, VTableRow<S> row, VTableSharedData<S> shared) {
        this.row = row;
        if (node != null) {
            getChildren().add(node);
        }
        setMinHeight(30);
        setAlignment(Pos.CENTER_LEFT);
        setPadding(new Insets(2, 5, 2, 5));

        setOnMouseClicked(e -> shared.tableView.selectRow(row));
    }

    public void resetBg() {
        backgroundSetByUser = null;
        row.setBgColor();
    }

    void setBackground0(Background bg) {
        if (backgroundSetByUser != null) {
            setBackground(backgroundSetByUser);
        } else {
            setBackground(bg);
        }
    }

    public void setBg(Background bg) {
        backgroundSetByUser = bg;
        row.setBgColor();
    }
}

File: src-ts/src/ui/table/VTableCellPane.ts
import { StackPane } from '../../javafx/Pane.js';
import { Node } from '../../javafx/Node.js';
import { Insets, Pos, Background } from '../../javafx/layout.js';
import { posToFlex } from '../../javafx/layout.js';
import type { ChangeListener } from '../../javafx/Property.js';
import type { VTableRow } from './VTableRow.js';
import type { VTableSharedData } from './VTableSharedData.js';

export class VTableCellPane<S> extends StackPane {
  private readonly row: VTableRow<S>;
  readonly heightWatcher: ChangeListener<number> = (_old, now) => {
    if (now === null) return;
    const dNow = now;
    const thisHeight = this.getHeight();
    if (thisHeight < dNow) {
      this.setPrefHeight(dNow);
    }
  };
  private backgroundSetByUser: Background | null = null;

  constructor(node: Node | null, row: VTableRow<S>, shared: VTableSharedData<S>) {
    super();
    this.row = row;
    if (node !== null) {
      this.getChildren().add(node);
    }
    this.setMinHeight(30);
    this.setAlignment(Pos.CENTER_LEFT);
    this.setPadding(new Insets(2, 5, 2, 5));

    this.setOnMouseClicked(() => shared.tableView.selectRow(this.row));
  }

  // TS StackPane is a flex container, so Pos is translated into
  // align-items/justify-content.
  setAlignment(p: Pos): void {
    const { justify, align } = posToFlex(p);
    this.el.style.alignItems = align;
    this.el.style.justifyContent = justify;
  }

  resetBg(): void {
    this.backgroundSetByUser = null;
    this.row.setBgColor();
  }

  setBackground0(bg: Background): void {
    if (this.backgroundSetByUser !== null) {
      this.setBackground(this.backgroundSetByUser);
    } else {
      this.setBackground(bg);
    }
  }

  setBg(bg: Background): void {
    this.backgroundSetByUser = bg;
    this.row.setBgColor();
  }
}
